uaclient/files.py

```diff
@@ -0,0 +1,296 @@
+import json
+import logging
+import os
+from datetime import datetime
+from typing import Dict, Optional
+
+from uaclient import defaults, event_logger, schemas, util
+
+PRIVATE_SUBDIR = "private"
+
+event = event_logger.get_event_logger()
+LOG = logging.getLogger(__name__)
+
+
+class UAFile:
+    _root_directory = defaults.DEFAULT_DATA_DIR
+    _file_name = "file"
+
+    def __init__(
+        self,
+        root_directory: Optional[str] = None,
+        filename: Optional[str] = None,
+        private: bool = True,
+    ):
+        if root_directory:
+            self._root_directory = root_directory
+        if filename:
+            self._file_name = filename
+        self._is_private = private
+        self._filepath = None
+
+    @property
+    def filepath(self):
+        if not self._filepath:
+            self._filepath = os.path.join(
+                self._root_directory, self._file_name
+            )
+        return self._filepath
+
+    @property
+    def is_private_file(self):
+        return self._is_private
+
+    def write(self, content):
+        file_mode = 0o600 if self._is_private else 0o644
+        self.create_file()
+        if not isinstance(content, str):
+            content = json.dumps(content, cls=util.DatetimeAwareJSONEncoder)
+        util.write_file(self.filepath, content, file_mode)
+
+    def read(self):
+        content = None
+        try:
+            content = util.load_file(self.filepath)
+        except FileNotFoundError:
+            if not os.path.exists(self.filepath):
+                LOG.debug("File does not exist: {}".format(self.filepath))
+        return content
+
+    def delete(self):
+        util.remove_file(self.filepath)
+
+    def create_file(self):
+        file_dir = os.path.dirname(self.filepath)
+        if not os.path.exists(file_dir):
+            os.makedirs(file_dir)
+            if self._is_private:
+                os.chmod(file_dir, 0o700)
+
+
+class PublicMachineTokenFile(UAFile):
+    _file_name = defaults.PUBLIC_MACHINE_TOKEN_FILE
+    _is_private = False
```

Comment:
I don't think these need to be class properties. they can just be instance properties that are set in the constructor